sycl/doc/extensions/experimental/sycl_ext_oneapi_usm_raii.asciidoc

```diff
@@ -0,0 +1,135 @@
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2025 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 9 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is an experimental extension specification, intended to provide early
+access to features and gather community feedback.  Interfaces defined in this
+specification are implemented in {dpcpp}, but they are not finalized and may
+change incompatibly in future versions of {dpcpp} without prior notice.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Overview
+
+SYCL 2020 provides very C-like interfaces for USM allocation/deallocations
+routines. With in-order queues/USM and
+link:./sycl_ext_oneapi_enqueue_functions.asciidoc[sycl_ext_oneapi_enqueue_functions]
+as a future direction to write SYCL programs with as little overhead as possible
+we need usability improvements for the USM APIs that C++ programmers would
+expect.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_USM_RAII` to one of the values defined in
+the table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== USM Deleter
+
+This extension adds `usm_deleter` helper that can be used with `std::unique_ptr`
+or `std::shared_ptr` to manage deallocation of USM memory.
+
+[source, c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+class usm_deleter {
+public:
+  usm_deleter(const context &Ctx);
+  usm_deleter(const queue &q);
+
+  template <typename T> void operator()(T *Ptr) const;
+};
+} // sycl::ext::oneapi::experimental
+----
+
+=== Example usage
+
+A simple example of using this extension is below.
+
+```c++
+{
+  sycl::queue q;
+  using sycl::ext::oneapi::experimental::usm_deleter;
+  std::unique_ptr<int, usm_deleter> p{sycl::malloc_shared<int>(1, q), {q}};
+  q.single_task([=, p = p.get()]() { *p = 42; }).wait();
```

Comment:
What would happen if the user hadn't called `wait` here?

I've never really used `shared_ptr` or `unique_ptr` before, but I'm a little worried that automatically invoking delete when the object is destructed could very easily lead to bugs (e.g., if a USM allocation is freed while a command is still running on the GPU). This might be fine, but it might be worth calling out in a non-normative note so developers are aware of the potential issues.

Reply:
That isn't different from doing something like
 
```
{
  auto p = std::make_unique<...>();
  std::thread t{[p = p.get()]() { use(p); }};
  // Forgotten: t.join();
}
```

which would be UB as well, so it shouldn't be too surprising for people used to using these smart pointer classes. That said, a small note won't hurt, so I'll add it.

Reply:
We could have the `usm_deleter` store a shared_ptr to the `queue_impl` rather than the `context_impl`. And then, before calling `free()` it could call `wait()`  .   That would solve this problem.  Though, I'm not advocating that we do that.

It woudl be at the cost of introducing too much stall.  We already have an issue where USM is slower than buffers when doing lots of creation/deletion, because some USM operations implicitly wait, introducing needless stalls. So, probably we shouldn't exacerbate it.